lib/util/ts_parser.js

```diff
@@ -343,21 +343,32 @@ shaka.util.TsParser = class {
    * @return {!Array.<shaka.extern.ID3Metadata>}
    */
   getMetadata() {
+    const timescale = shaka.util.TsParser.Timescale_;
+    const Uint8ArrayUtils = shaka.util.Uint8ArrayUtils;
     const metadata = [];
-    let id3Data = new Uint8Array(0);
-    for (const data of this.id3Data_) {
+    let prevId3Data = new Uint8Array(0);
+    // parsePES() only works if the data begins on a PES boundary.
+    // Try the last data blob first, and if it doesn't begin on a
+    // PES boundary, prepend the previous blob and try again.
+    // This way, a successful parse will always begin and end on
+    // the correct boundary, and no data will be skipped.
```

Comment:
The only thing I still don't understand is why this is better/different from concatenating everything together and calling parse exactly once.

Reply:
Because each PES has a different PTS

Reply:
I see.  So we the contents of this.id3Data_ are split at arbitrary places, and we really only want to parse one PES at a time.  Is that correct?

Reply:
It’s correct!

Reply:
If they are split at arbitrary places, this could devolve into the exact case we had before: one buffer with everything in it.  In that case, it might be best to change parsePES instead to report how much data it consumed, so that the leftovers can roll forward.  We do something similar to that for LL-DASH, since streaming network buffers are arbitrary amounts of data, and we need to regroup it into mp4 boxes.

Reply:
sure, but in mp4 the boxes include everything, and here several packages are necessary to make up a PES. I'm not sure if I understand what you want...

Reply:
Here's what I'm thinking.  If a PES packet were "ABCDE", all PES packets began with "A", and the data array might look like this:

[
  "ABC",
  "DEA",
  "BCD",
  "EAB",
  "CDE",
]

That's 3 packets broken into 5 pieces.  With the way this is written right now, you'll end up combining them into a single buffer:

"CDE" ... doesn't start with "A", not a PES packet.
"EABCDE" ... same
"BCDEABCDE" ... same
"DEABCDEABCDE" ... same
"ABCDEABCDEABCDE" ... finally parsed.

The result is the same as what you replaced.

Will that never happen?  For example, I could imagine your data could always be broken up like this:

[
  "ABC",
  "DE",
  "ABC",
  "DE",
  "ABC",
  "DE",
]

If that were always the way the data array was broken up, your change would make sense as is.  Going backwards means you always have exactly one complete packet when the combined buffer starts with "A".

If the first scenario is possible, then you could do what we do for mp4 parsing w/ ReadableStream.  The parser tracks how much data was and wasn't consumed, and it goes forward.  For example:

[
  "ABC",
  "DEA",
  "BCD",
  "EAB",
  "CDE",
]

"" => "ABC" ... incomplete.  0 outputs, 3 bytes left: "ABC"
"ABC" => "ABCDEA" ... found a packet.  1 output, 1 byte left: "A"
"A" => "ABCD" ... incomplete.  0 outputs, 4 bytes left: "ABCD"
"ABCDEAB" ... found a packet.  1 output, 2 bytes left: "AB"
"AB" => "ABCDE" ... found a packet.  1 output, 0 bytes left.

Reply:
As far as I understand and with all the streams I've tried (many these days), it can only be the case of:
[
"ABC",
"DE",
"ABC",
"DE",
"ABC",
"DE",
]

So I think this change is fine.